internal/parser/parser.go

```diff
@@ -1049,6 +1055,79 @@ func (p *parser) repeat(fn func() bool) {
 	}
 }
 
+// open will open a new block. It will expect that the next token
+// is the start token and mark that we expect the end token in the
+// future.
+func (p *parser) open(start, end token.Token) (pos token.Pos, lit string) {
+	pos, lit = p.expect(start)
+	p.stack[end]++
+	return pos, lit
+}
+
+// more will check if we should continue reading tokens for the
+// current block. This is true when the next token is not EOF and
+// the next token is also not one that would close a block.
+func (p *parser) more() bool {
+	_, tok, _ := p.peek()
+	if tok == token.EOF {
+		return false
+	}
+	_, ok := p.stack[tok]
+	return !ok
+}
+
+// close will close a block that was opened using open.
+//
+// This function will always decrement the stack for the end token
+// and it will remove the entry if the stack reaches zero for that
+// token.
+//
+// If the next token is the end token, then this will consume the
+// token and return the pos and lit for the token. Otherwise, it will
+// return NoPos.
+// TODO(jsternberg): NoPos doesn't exist yet so this will return the
+// values for the next token even if it isn't consumed.
+func (p *parser) close(end token.Token) (pos token.Pos, lit string) {
+	// If the end token is EOF, we have to do this specially
+	// since we don't track EOF.
+	if end == token.EOF {
+		// TODO(jsternberg): Check for EOF and panic if it isn't.
+		pos, _, lit := p.scan()
+		return pos, lit
+	}
+
+	// The end token must be in the stack. We should have already
+	// checked for this.
+	count, ok := p.stack[end]
+	if !ok {
+		panic("closing a block that was never opened")
+	}
+
+	// Decrement the stack whether or not the next token is the
+	// end token.
+	if count == 1 {
+		delete(p.stack, end)
```

Comment:
With my comment above you could leave this token in the map and remove this branching logic. I.e always decrement.